File: destack-py/destack/language/core/runtime/connection.py
from collections.abc import Sequence
from typing import TYPE_CHECKING

from destack.language.core import (
    Handle,
    HandleType,
    NodeReference,
    builtin_handle,
    builtin_property_runtime,
)
from destack.utils.uuid import UUID

if TYPE_CHECKING:
    from destack.language import Event


@builtin_handle(HandleType.CONNECTION)
class Connection(Handle):
    """
    A connection between a local and a remote Graph.
    """

    space_ptr: "NodeReference" = builtin_property_runtime(500, is_repr=True)

    async def open(self) -> None:
        raise NotImplementedError

    async def pull(
        self,
        space_id: UUID,
        branch_id: UUID | None,
        snapshot_id: UUID | None,
    ) -> None:
        """Pull the relevant Entities and Events from the remote Graph into this Graph."""
        raise NotImplementedError

    async def push(self, events: Sequence["Event"]) -> Sequence["Event"]:
        """Push the Events to the remote Graph."""
        raise NotImplementedError

    async def close(self) -> None:
        raise NotImplementedError

File: destack-ts/src/language/core/runtime/connection.ts
import type { Event, Graph, NodeReference, Session } from "@destack/language";

/**
 * A connection between a local and a remote Graph.
 */
export class Connection {
  /** The Session this Connection is in. */
  readonly session: Session;

  /** The local Graph. */
  readonly graph: Graph;

  /** The remote Space. */
  readonly remoteSpaceRef: NodeReference;

  constructor(options: { remoteSpaceRef: NodeReference; graph: Graph; session: Session }) {
    this.remoteSpaceRef = options.remoteSpaceRef;
    this.session = options.session;
    this.graph = options.graph;
  }

  repr(): string {
    return `<Connection remote=${this.remoteSpaceRef.id}>`;
  }

  async open(): Promise<void> {
    throw new Error("not implemented");
  }

  async commit(events: Event[]): Promise<Event[]> {
    throw new Error("not implemented");
  }

  async close(): Promise<void> {
    throw new Error("not implemented");
  }
}
